### ISSUE TITLE
Support for ASP.NET Core 3.0

### ISSUE THREAD

**Issue by robertmircea** [castleproject/Windsor#493]
What is the recommended way to use Castle Windsor with asp.net core 3.0? One place where it can integrated would be the generic host towards which asp.net core 3.0 migrated, probably using `UseServiceProviderFactory`

Using `Castle.AspnetCore.Facility` is no longer possible (or at least I didn't found out how) because Startup.ConfigureServices returning IServiceProvider is no longer supported by the framework and AddWindsor cannot be used anymore. Before I used to write something like this in Startup class

```
public IServiceProvider ConfigureServices(IServiceCollection services)
{
    container.AddFacility<AspNetCoreFacility>(f => f.CrossWiresInto(services));
[...]
    return services.AddWindsor(container, 
        opts =>
         {
            opts.RegisterControllers(typeof(MyController).Assembly, LifestyleType.Transient);
         });
}
```

**Comment by jonorossi** [comment]
Probably worth getting together with @schwarzie2478 in #487.

**Comment by tasoss** [comment]
I have the same problem.
I just don't understand if there's going to be a solution soon.
Can you please let me know?
Thanks!

**Comment by robertmircea** [comment]
As it seems it looks like a dead subject unfortunately since I don't think that AspNetCoreFacility has any active maintainers.

I've managed to make it work in a .net core 3.0 app by insourcing in my project some of the classes from ASP.NET Core Facility and modifying and eliminating some stuff which I did not need like Tag helpers support whose public APIs changed in the new .net core release. The code was in `WindsorRegistrationHelpers.AddApplicationComponentsToWindsor` called by extension method named AddWindsor. My app is providing only web apis so fortunately I needed to register only controllers with Windsor.

**Comment by jonorossi** [comment]
> since I don't think that AspNetCoreFacility has any active maintainers.

@robertmircea correct, no active maintainers as I said in #487.

**Comment by tasoss** [comment]
Thank you for the information

**Comment by River39** [comment]
I was gone use Windsor in my core 3 application until i found this. Has any progress been made on this issue? Or are there alternative solutions to integrate windosr in core 3?

**Comment by dotnetjunkie** [comment]
Here are two issues that I filed with Microsoft concerning integrating Simple Injector with ASP.NET Core 3:

* [Microsoft.Extensions.Hosting.Host creates hosted services too early](https://github.com/aspnet/AspNetCore/issues/14585)
* [How to integrate non-conforming containers with Razor Components?](https://github.com/aspnet/AspNetCore/issues/8886)

I'm pinging you about this, because it might have impact on integrating Castle Windsor with ASP.NET Core 3 as well.

**Comment by chrisheil** [comment]
If I am reading the comments in this issue correctly, there is no official support for ASP.NET Core 3.0, and no plans to add support in the future. Is that correct?

**Comment by jonorossi** [comment]
@chrisheil Windsor today does not have first party support for ASP.NET Core 3.0, correct.

It isn't that there are no plans to add support in the future, it is just that no one has contributed it yet. Both @stakx and myself are the only active contributors to all Castle projects, we are individual contributors like most contributors have been over a long time, we are not being paid for work on Castle. Unfortunately without a contributor that is interested in both Windsor and ASP.NET Core stepping up it just isn't going to magically happen, there have been a bunch of people that have shown interest, and I've offered to review designs and code, but there has been no movement.

**Comment by marcelNgan** [comment]
> As it seems it looks like a dead subject unfortunately since I don't think that AspNetCoreFacility has any active maintainers.
> 
> I've managed to make it work in a .net core 3.0 app by insourcing in my project some of the classes from ASP.NET Core Facility and modifying and eliminating some stuff which I did not need like Tag helpers support whose public APIs changed in the new .net core release. The code was in `WindsorRegistrationHelpers.AddApplicationComponentsToWindsor` called by extension method named AddWindsor. My app is providing only web apis so fortunately I needed to register only controllers with Windsor.

Hi @robertmircea , also trying to use Windsor here with ASP.Net Core 3.0 after using Windsor for a few years with .Net Framework. It's my first project with Core so I'm a bit confused here as well. I'm only trying to use my app for WebApis so potentially this might still work? Any chance you can offer some guidance of what you are doing?

**Comment by chrisheil** [comment]
@jonorossi Thank you for the response. I appreciate the effort your and everyone else has put into this project, and understand the situation. Would it be possible to add this disclaimer to the AspNetCore documentation to make this a little more clear to anyone else looking for this information? There is a page demonstrating how to set up in an asp.net core 2.0 environment which led to a ton of wasted time trying (and failing) before reaching this conclusion. I could create a PR for this update as well if you would prefer

@marcelNgan  I have a workaround that is a little flawed, but is "workable"

Take a look at [Volosoft's](https://github.com/volosoft/castle-windsor-ms-adapter) adapter. This in your Program.cs file add the following lines

`return Host.CreateDefaultBuilder(args)
                .UseServiceProviderFactory(new WindsorServiceProviderFactory())
                .ConfigureContainer<WindsorContainer>((hostBuilderContext, windsorContainer) =>
                {

                    windsorContainer.Kernel.Resolver.AddSubResolver(new CollectionResolver(windsorContainer.Kernel, true));

                    windsorContainer.AddFacility<TypedFactoryFacility>();
                    windsorContainer.Install(
                        FromAssembly.InThisApplication(typeof(Program).Assembly),
                        FromAssembly.Instance(typeof(SomeOtherClassInAnotherAssembly).Assembly)
                    );
                })`

This seems to be working on my initial tests. The weird behavior I noticed is the same as described [in this pull request](https://github.com/volosoft/castle-windsor-ms-adapter/issues/33) where the adapter isn't properly disposing when trying to Ctrl+C out of "dotnet run" in a terminal.

The project itself was [proposed to be included in Castle Winsdor](https://github.com/volosoft/castle-windsor-ms-adapter/issues/28) itself, but I wasn't able to find documentation to see how the new version is working. The Volosoft project also does not look to be actively maintained anymore, which is also risky.

**Comment by jonorossi** [comment]
@chrisheil happy to accept a PR to add a disclaimer to the docs that the current facility does not work with ASP.NET Core 3.0.

**Comment by dquist** [comment]
Fellas this is an interesting topic!

In my scenario I am working a greenfield WPF project and was looking to leverage some of the new hosting features offered in .net core, specifically the configuration and logging management. However I wanted to stick with Windsor as an IoC container because of how capable it is.

I came up with a solution that I realize may not be the most orthodox, but it seems to do the trick. Basically I use the `HostBuilder` to to bootstrap a hosted service that contains the Windsor application. I then created a sub-resolver for Windsor that will cross-wire all the Microsoft provided loggers to the Windsor application. This way I can get the best of both worlds, my Windsor IoC container and all the logging goodies.

Here's my Windsor application implemented as a hosted service. The `ILoggerFactory` dependency is resolved by the .net DI container and is used to cross-wire the `ILogger` instances.

```c#
public sealed class HostedWindsorApplication : IHostedService, IDisposable
{
    private readonly ILoggerFactory _loggerFactory;
    private readonly IWindsorContainer _container;

    public HostedWindsorApplication(ILoggerFactory loggerFactory)
    {
        _loggerFactory = loggerFactory ?? throw new ArgumentNullException(nameof(loggerFactory));
        _container = new WindsorContainer();
    }

    public async Task StartAsync(CancellationToken cancellationToken)
    {
        // Use sub-resolver to cross-wire ILogger instances
        _container.Kernel.Resolver.AddSubResolver(new LoggerResolver(_loggerFactory));

        _container.Install(new CoreInstaller());
        var app = _container.Resolve<IMyApplication>();
        await app.Run().ConfigureAwait(false);
        _container.Release(app);
    }

    public Task StopAsync(CancellationToken cancellationToken)
    {
        return Task.CompletedTask;
    }

    public void Dispose()
    {
        _container?.Dispose();
    }
}
```

Then my App.xaml.cs

```c#
public sealed partial class App
{
    private IHost? _host;

    protected override async void OnStartup(StartupEventArgs e)
    {
        _host = new HostBuilder()
            .ConfigureHostConfiguration(c =>
            {
                c.SetBasePath(Directory.GetCurrentDirectory());
                c.AddJsonFile("appsettings.json", optional: true, reloadOnChange: true);
                c.AddCommandLine(e.Args);
            })
            .ConfigureServices((context, services) =>
            {
                // Use the built-in DI to bootstrap the windsor application
                services.AddHostedService<HostedWindsorApplication>();
            })
            .ConfigureLogging((context, builder) =>
            {
                builder
                    .ClearProviders()
                    .AddConfiguration(context.Configuration)
                    .SetMinimumLevel(LogLevel.Trace)
                    .AddNLog("NLog.config")
                    .AddSentry(o =>
                    {
                        o.ConfigureScope(s => s.SetTag("RootScope", "sent with all events"));
                    });
            })
            .Build();
        await _host.StartAsync().ConfigureAwait(false);
    }

    protected override async void OnExit(ExitEventArgs e)
    {
        using (_host)
        {
            await _host.StopAsync(TimeSpan.FromSeconds(5)).ConfigureAwait(false);
        }
    }
}
```

And finally my sub-resolver to cross-wire the `ILogger` instances. This one was kind of tricky since the loggers can be declared as `ILogger<ResolvingType>`. I had to convert the `ILogger<>` type into a generic type based on the handler type, and then create a new instance using Activator. There may be a better way to handle this that I'm not aware of.

```c#
public class LoggerResolver : ISubDependencyResolver
{
    private readonly ILoggerFactory _loggerFactory;

    public LoggerResolver(ILoggerFactory loggerFactory)
    {
        _loggerFactory = loggerFactory ?? throw new ArgumentNullException(nameof(loggerFactory));
    }

    public bool CanResolve(CreationContext context, ISubDependencyResolver contextHandlerResolver, ComponentModel model, DependencyModel dependency)
    {
        return typeof(ILogger).IsAssignableFrom(dependency.TargetType);
    }

    public object Resolve(CreationContext context, ISubDependencyResolver contextHandlerResolver, ComponentModel model, DependencyModel dependency)
    {
        // Resolve loggers of type ILogger<>
        if (dependency.TargetType.IsGenericType && dependency.TargetType.GetGenericTypeDefinition() == typeof(ILogger<>))
        {
            var genericLoggerType = typeof(Logger<>).MakeGenericType(context.Handler.ComponentModel.Implementation);
            return Activator.CreateInstance(genericLoggerType, _loggerFactory);
        }

        // Resolve pure ILogger types
        return _loggerFactory.CreateLogger(model.Implementation.FullName);
    }
}
```

That's what I've got working right now for my WPF application. I've only hammered it out in the past couple days so it's probably still a bit rough around the edges but it seems to be doing the trick.

**Comment by generik0** [comment]
Hi All.

I helped @fir3pho3nixx with the original aspnetcore2.x implementation (@fir3pho3nixx did most of the work, this was the way he wanted it).

Does anyone know yet, if we are in a start, to make a 3.x facility for AspNetCore. I will be needing something because EF Core 3.x requires netapp3.x 
I can start up in march on looking at this again.

Before that, is anyone interested in grouping up with me, and figuring this out?
We also need a solution for the generic host.

**Comment by rvdginste** [comment]
@dotnetjunkie 

From: https://github.com/dotnet/aspnetcore/issues/14585
> During my research I found a way to circumvent the behavioral difference between
> Host and WebHost in a way that doesn't require the use of IServiceProviderFactory<T>.

Can you tell something more about that?

@generik0 

Do you still plan on working on this?

**Comment by generik0** [comment]
Hi @dotnetjunkie.
Yes, thanks. I want to give it a shot. I worked a bit with Gavin on the aspnetcore 2 facility, where we stole your cross registering, so I *hope* I can work it out.

I am a little worried that nobody else has come forward with intentions for assistance, makes me a little worried that the facility is not needed.

But please, and links/advice you have are more than welcome!!!

**Comment by dotnetjunkie** [comment]
@rvdginste,

Let me first summarize the problem with ASP.NET Core 3:

> ASP.NET Core 3.0's new Microsoft.Extensions.Hosting.Host creates hosted services between Startup's ConfigureServices and Configure methods, where the 'old' Microsoft.AspNetCore.WebHost creates the hosted services only after the Configure method has run. This behavior difference between Microsoft.AspNetCore.WebHost and Microsoft.Extensions.Hosting.Host is significant, because it disallows users of non-conforming containers (like Simple Injector, Castle Windsor, and Ninject) to resolve hosted services from their container, because the configuration of those containers needs to be finished in the Startup.Configure phase, while resolving hosted services from that container before that configuration is finished, can lead to problems. Simple Injector, for instance, blocks any registrations made after the first resolve (which will be a resolve for MyHostedService if you assume the hosted service to be added using `p => container.GetInstance<MyHostedService>())`. But even if the container doesn't block registrations, the early request for hosted services can cause trouble, because the hosted service's dependencies might not be registered at that point.

The way we were able to circumvent this problem is by adding an ` ContainerLocking` event to the container that signals when the container is about to get locked. This happens, for instance, when the first call to `GetInstance` is being made, but just before the instance is being resolved and before the container is being locked. The ASP.NET Core integration packages hook onto that event in order to do their last-minute registrations. This works around the problem because all registrations need to (obviously) be done before the container gets locked (which is when the first service is being requested), the first resolve might happen between `ConfigureServices` and `Configure` in the case of the new `Host`, while it would always happen after `Configure` in the old model.

Perhaps unsurprisingly, but this change was a time-consuming endeavor. We needed a lot of discussions, developing, testing, and documenting to get this fix out the door.

**Comment by generik0** [comment]
@dotnetjunkie I like your point in https://github.com/dotnet/aspnetcore/issues/14585#issuecomment-536742102.

Personally, I wish you had kept on the discussion with @davidfowl regarding the IServiceProviderFactory and aspnetcore3.0 issues with non-conforming DI. I would have loved to give you any backing I could (but I doubt you needed).

My thoughts are, once again, if we want the aspnetcore facility to work in Windsor, we need to look to simple injector for guidance.

**Comment by rvdginste** [comment]
@dotnetjunkie Thank you for the explanation!

@generik0 I am willing to help in whatever way I can, but personally I am only interested in support for controllers. So I made a test project that includes the existing facility code to see how the current facility works (using asp.net core 3 with the old WebHost), but I threw out the code related to TagHelpers and ViewComponents.

**Comment by davidfowl** [comment]
I think this is the cleanest way to plug in both a conforming and non-conforming container into the system:
https://github.com/dotnet/aspnetcore/issues/14585#issuecomment-536930119

It's a way to give you a a chance to create your own builder (whatever that may be) and it gives you a chance to return a service provider (it a clean way to plug into the lifecycle of the system).

**Comment by generik0** [comment]
@rvdginste that sounds great! can you share it in e github repo?

**Comment by generik0** [comment]
HI @davidfowl ! Thanks for joining the conversation!! :-)
I think I understand your point about 14585. It just looks strange to me that we essentially need to double register "stuff". i.e first with our own DI  (conforming or not) and then the with ServiceCollection . We got around this in aspnetcore2.0 with cross registering, 

My head doesn't quite understand why MS didn't make the Service collection etc, replaceable with any DI. Why would you lock it to your specific implementation? Or am I wrong here? Is the MS DI replaceable with anything, just by implementing some simple interfaces on our own implementations? And registering some standard MS implementations in "our" DI?

The MS Service Collection is probably designed and programmed very will. However the other DIs like Caste and Simple, just  have so many more powerful features, that we need. they are years in front of the inbuild Di..

Why does MS not provide a simple way for DIs to integrate into the new hosts? One would think it was to your benefit? Or have I misunderstood something, is it very easy or override e.g. Simple and Castle in the new net core release? The example you provide does not seam to make it easy.

Also this whole conforming and none conforming questions. Why shouldn't we be able to register stuff? Just because Martin Fawler said so? Are we really going to make architecture based upon a principle someone made 20 years ago (whichj might have been incorrect since Simple, Castle and Ninject have many use cases where something else was needed). Can we never rethink architecture/prinicples?

**Comment by dquist** [comment]
Can confirm that using @davidfowl's approach works. I've successfully used this approach with
https://github.com/volosoft/castle-windsor-ms-adapter

to 'cross-wire' the DI registrations.

It is sightly unintuitive because the services registered in `ConfigureServices` aren't actually registered with Windsor - they're registered to the `IServiceCollection` and then re-registered with Windsor afterwards, but it is doable.

**Comment by robsonj** [comment]
We have a large suite of code that utilizes Castle.Windsor. We looking to move our web framework over from being NancyFx based to ASP.Net Core. I'm wondering if it makes sense to start this with the WebHost still given the issues I've read through above? It's not completely clear to me (due to my nativity with ASP.Net Core) if there is any way to bridge the gap between asp.net core and Castle.Windsor yet? We're using Castle.Windsor v5.0.1

**Comment by chrisheil** [comment]
@robsonj I made a similar migration last year, only the added joy of going from WCF to asp.net core endpoints. The suggestions above are working just fine for us now, we have not encountered any issues yet related to Castle.Windsor on dotnet core 3.

Outside of figuring out the appropriate Lifestyles of course

**Comment by robsonj** [comment]
> @robsonj I made a similar migration last year, only the added joy of going from WCF to asp.net core endpoints. The suggestions above are working just fine for us now, we have not encountered any issues yet related to Castle.Windsor on dotnet core 3.
> 
> Outside of figuring out the appropriate Lifestyles of course

My confusion was with regard to which solution to utilize from the conversation above, looks like a couple of solutions in this thread, along with some other referenced solutions for other DI frameworks, so I wasn’t sure which way to go. I’m new to asp.net core , so pardon my ignorance etc

**Comment by chrisheil** [comment]
@robsonj I would recommend trying the solution dquist provided [here](https://github.com/castleproject/Windsor/issues/493#issuecomment-558910274) first

If that does not work for you, I can verify [my solution](https://github.com/castleproject/Windsor/issues/493#issuecomment-547450198) is usable for now, but with the issue I linked to above.

**Comment by robsonj** [comment]
> @robsonj I would recommend trying the solution dquist provided [here](https://github.com/castleproject/Windsor/issues/493#issuecomment-558910274) first
> 
> If that does not work for you, I can verify [my solution](https://github.com/castleproject/Windsor/issues/493#issuecomment-547450198) is usable for now, but with the issue I linked to above.

Thank you, I’ll give it a shot, thanks for the help!

**Comment by robsonj** [comment]
Ok, got a bit further. I followed the example above, but I was expecting the Microsoft.Extensions.Logging.ILogger<> dependency in my Razor page .cs (Index.chtml.cs) to result in a call on the LoggerResolver CanResolve() being invoke to determine its actual implementation, that doesn't seem to be the case though? I also tried having the Razor page depend on the Castle ILogger, but that just results in a runtime error. 

I'm I doing something wrong, or are my expectations incorrect?

Cheers
Jonathan

PS Let me know if this is the wrong place to be asking these questions.

**Comment by dquist** [comment]
@robsonj My initial implementation above is pretty rough and I wouldn't recommend it. It uses two separate containers, the Microsoft provided one and Windsor. The sub-resolver allows the Windsor container to resolve instances from the Microsoft container, but not the other way around. That's the biggest drawback is that you have to depend on the Windsor container itself.

I ended up using the https://github.com/volosoft/castle-windsor-ms-adapter project which works well enough and follows Microsofts design for an ASP.NET Core container.

Hope that helps.

**Comment by robsonj** [comment]
> @robsonj My initial implementation above is pretty rough and I wouldn't recommend it. It uses two separate containers, the Microsoft provided one and Windsor. The sub-resolver allows the Windsor container to resolve instances from the Microsoft container, but not the other way around. That's the biggest drawback is that you have to depend on the Windsor container itself.
> 
> I ended up using the https://github.com/volosoft/castle-windsor-ms-adapter project which works well enough and follows Microsofts design for an ASP.NET Core container.
> 
> Hope that helps.

Thank you, I was just looking at the volosoft adapter. Thanks for confirming my understanding was incorrect. We definitely wanna use the Castle ILogger everywhere if possible as we have an entire code base that depends on it (log4net under the covers)

**Comment by robsonj** [comment]
Ok, the volosoft route seems to work better. Got it working much better in 22 minutes :-). Thank you for the help all!

**Comment by ltines** [comment]
Hi,

had a stab at it, check out https://github.com/EurofficeGroup/windsor-service-provider . It took few days to understand al the .NET Core interfaces. Massive thanks to volosoft for figuring out bunch of the semantics.

This approach, similar to volosoft, doesn't try to run the default container (`DefaultServiceProvider`) and Windsor side by side but rather using Windsor as the one and only container. It supports .NET Core scopes as well.

There are few things I am not sure about:
* what's the semantics of Transient in .NET Core (i.e. `services.AddTransient`)
* tried it only with WebAPI project. Other features might not work but happy to collaborate on getting them working

At some point might be useful to run it through [Microsoft.Extensions.DependencyInjection.Specification.Tests](https://www.nuget.org/packages/Microsoft.Extensions.DependencyInjection.Specification.Tests)

**Comment by davidfowl** [comment]
> At some point might be useful to run it through Microsoft.Extensions.DependencyInjection.Specification.Tests

That's really the most important part if you go this route. Any failing test will result in some incompatibility in behavior.

**Comment by generik0** [comment]
@ltines very cool,very interesting!
Are you proposing we replace the castle.aspnetcore. Facility with this implementation?

**Comment by ltines** [comment]
> 
> 
> > At some point might be useful to run it through Microsoft.Extensions.DependencyInjection.Specification.Tests
> 
> That's really the most important part if you go this route. Any failing test will result in some incompatibility in behavior.

Yup, gave it a shot - have 5 tests failing. Mostly around scopes semantics.



> 
> 
> @ltines very cool,very interesting!
> Are you proposing we replace the castle.aspnetcore. Facility with this implementation?

IMHO it's cleaner solution than having two containers side-by-side

**Comment by generik0** [comment]
@ltines Yes, if you/we can get the MS "visioned" implementation of the DI to work using one the Windsor, I think it would be magnificent!
I noticed you Apache License 2.0'ed you repo. Your ok to move to Castle?

@jonorossi:
- How do you feel about all this. Should there be branched, and reworked the current castle.aspnetcore.facility = ?
- We also need the net app generic host to get working, so it isn't "just" aspnetcore. is castle.aspnetcore.factility the correct name?

I personally promised to also start looking at this in March. Unfortunately, I got one month behind on a major project and managed to get bronchitis (in this time of Corona) {excuses, excuses} but @ltines do you want any help, i think i can manage some time?

**Comment by ltines** [comment]
> @ltines Yes, if you/we can get the MS "visioned" implementation of the DI to work using one the Windsor, I think it would be magnificent!
> I noticed you Apache License 2.0'ed you repo. Your ok to move to Castle?
> 
Sure, no problem with that.

I am down to last the failing test - ordering of scoped components disposal. 

@generik0 
* there is expectation that Transient components are actually tracked and disposed on scope dispose. So I made them scoped (while still returning new instance on Resolve). Not sure if it's the right (tm) solution. Opinions?
* would be good if you have any reasonably complex Razor / WebPages project to give it a shot.

**Comment by davidfowl** [comment]
I'll chime in:

> We also need the net app generic host to get working, so it isn't "just" aspnetcore. is castle.aspnetcore.factility the correct name?

ASP.NET Core is not the correct name 😄 . See https://www.nuget.org/packages/Autofac.Extensions.DependencyInjection/

> would be good if you have any reasonably complex Razor / WebPages project to give it a shot.

Once those tests pass everything will work like magic

**Comment by generik0** [comment]
Thanks @davidfowl :-) I think wse need to wait for @jonorossi to chim in with naming. But there is an issue with the name. There is also some web scoping in the facility that nancyfx 2.0 needs. So maybe some things should be refactored :-)

@ltines I have an extremely complex, over engineered :-), angular, node services, entity FW project to integrate (with 1 razor/mvc page (login)). But yes, i think it would be very interesting to try on this. I also have a reasonably complex 100% mvc / razor + ts(webpack/npm build). The last is on net classic, so i need to see if i can migrate.

**Comment by generik0** [comment]
@ltines regardning the transient. I see the predicament. If the "parent object" is web scoped, then esentially all transients would also be web scoped (unless resolved from an injected factory). If the transient is injected into a singleton, it essentially is also singleton....
So in a way, a transient injected into a web scope, should also be disposed when the scope is disposed.

**Comment by generik0** [comment]
@ltines does the new and improved implementation work with the generic host?

**Comment by ltines** [comment]
@generik0 yeah the weird thing is though when you resolve Transient twice you should get two instances - so its something like ScopedTransient :)

And I believe it should work with Generic host - it uses `IHostBuilder` to replace `IServiceProviderFactory`

**Comment by generik0** [comment]
ctor (IGeneric0 generic1, IGeneric0 generic2)

**Comment by ltines** [comment]
@generik0 
when you resolve Transient component in Windosor, callee is responsible for releasing / disposing.
.NET Core has expectation that Transient components get disposed at the end of the scope or at disposing of ServiceProvider in root scope.

So Windsor Transient <> .NET Core Transient

**Comment by generik0** [comment]
@ltines i got this out somewhere in the current faciltiy. Used it for a nancy implementation. Does it help?

```c"
public class VemPerWebRequestLifestyleModule : IHttpModule
    {
        private const string Key = "castle.per-web-request-lifestyle-cache";

        private static bool initialized;

        public void Init(HttpApplication context)
        {
            initialized = true;
            context.EndRequest += EndRequest;
        }

        protected void EndRequest(Object sender, EventArgs e)
        {
            var scope = GetOrCreateScope(createIfNotPresent: false);
            scope?.Dispose();
        }

        internal static ILifetimeScope AttachScope()
        {
            EnsureInitialized();

            var context = HttpContext.Current;
            if (context == null)
            {
                throw new InvalidOperationException("HttpContext.Current is null. PerWebRequestLifestyle can only be used in ASP.Net");
            }

            return GetOrCreateScope(createIfNotPresent: true);
        }

        internal static ILifetimeScope DetachScope()
        {
            var scope = GetOrCreateScope(createIfNotPresent: true);
            if (scope != null)
            {
                HttpContext.Current.Items.Remove(Key);
            }

            return scope;
        }

        private static void EnsureInitialized()
        {
            if (initialized)
            {
                return;
            }

            throw new ComponentResolutionException($"PerWebRequestLifestyleModule was not initialised. Please refer to https://github.com/castleproject/Windsor/blob/master/docs/systemweb-facility.md for more info.");
        }

        private static ILifetimeScope GetOrCreateScope(bool createIfNotPresent)
        {
            var context = HttpContext.Current;
            if (context == null)
            {
                return null;
            }

            var candidates = (ILifetimeScope)context.Items[Key];
            if (candidates == null && createIfNotPresent)
            {
                candidates = new DefaultLifetimeScope(new ScopeCache());
                context.Items[Key] = candidates;
            }

            return candidates;
        }

        public void Dispose()
        {
        }
    }
```

**Comment by generik0** [comment]
@ltines and i guess the net core implementation doesn't have a nice release call, where it can dispose...

Can the current implememation be used? https://github.com/castleproject/Windsor/blob/4830718f99fd71e539133b34f9e54daa49049d6f/src/Castle.Facilities.AspNetCore/AspNetCoreExtensions.cs#L67

**Comment by generik0** [comment]
@ltines btw, a nice litttle mvc check for you is if these resolve / can be registered:
* IHttpContextAccessor
* IActionContextAccessor
* IHtmlLocalizer<SharedResource> (you need to make a resource file)
--> Component.For<IHtmlLocalizer<SharedResource>>().ImplementedBy<HtmlLocalizer<SharedResource>>()

:-)

**Comment by ltines** [comment]
> 
> 
> @ltines and i guess the net core implementation doesn't have a nice release call, where it can dispose...
> 
> Can the current implememation be used?
> 
> https://github.com/castleproject/Windsor/blob/4830718f99fd71e539133b34f9e54daa49049d6f/src/Castle.Facilities.AspNetCore/AspNetCoreExtensions.cs#L67

Actually it does. We're providing our own ScopeFactory and linking .NET Core scope with Windsor scope. I am just not perfectly happy with my solution.

**Comment by generik0** [comment]
Maybe this applies :-)
![image](https://user-images.githubusercontent.com/2411563/79212081-5dd15880-7e47-11ea-9423-261f77f5d2e3.png)

**Comment by jonorossi** [comment]
> How do you feel about all this. Should there be branched, and reworked the current castle.aspnetcore.facility = ?

@generik0 I'd leave the existing facility as is, its design is tied to ASP.NET Core which today is wrong, and we'll remove it as people move over, the new facility should not reference ASP.NET Core. I'd expect it to reference `Microsoft.Extensions.DependencyInjection` and `Microsoft.Extensions.Hosting` so you can use it in a console application without dragging in ASP.NET Core libraries.

> We also need the net app generic host to get working, so it isn't "just" aspnetcore. is castle.aspnetcore.factility the correct name?

@generik0 I agree with @davidfowl. I think the name should be `Castle.Facilities.MsDependencyInjection`.

**Comment by generik0** [comment]
@jonorossi That perfect. thx!!
Naming you mean: 
Castle.Facilities.Extensions.DependencyInjection
or
Castle.Windsor.Extensions.DependencyInjection

i think that is what @davidfowl refered to.. https://www.nuget.org/packages/Autofac.Extensions.DependencyInjection/

**Comment by jonorossi** [comment]
> Naming you mean:
> Castle.Facilities.Extensions.DependencyInjection
> or
> Castle.Windsor.Extensions.DependencyInjection

I don't see much value in the adding the "Extensions" part of the namespace, making it the full namespace would give you "Castle.Facilities.Microsoft.Extensions.DependencyInjection", which is silly.

For historic reasons all the first party facilities are under the facilities namespace, not the Windsor namespace.

I (not unlike the Volosoft guys) think that "Castle.Facilities.DependencyInjection" isn't clear enough this is for the .NET Core Extensions project support, therefore the extra prefix, DotNetExtensions as a prefix is just too long.

I thought exactly as I wrote `Castle.Facilities.MsDependencyInjection` made the most sense. If you disagree, happy to hear another proposal.

**Comment by jonorossi** [comment]
We are probably getting caught in the weeds here, but `Castle.Facilities.ExtensionsDependencyInjection` (without a dot) could also work?

I always see that Autofac example and think the `Autofac.Extensions.*` as a namespace for extensions (lowercase e), rather than `Extensions.DependencyInjection` as the integrated product.

**Comment by generik0** [comment]
@jonorossi I am "græskkatolsk" (Danish saying which means I don't mind either way). However, I do many times find naming harder than coding, and sometimes just as important getting right :-)

**Comment by ltines** [comment]
@generik0 @davidfowl 
Done few fixes and reworked how scoping works. Now it passes all the tests.

@generik0 @jonorossi 
I would suggest `Castle.Windsor.Extensions.DependencyInjection` mainly becase it's not a facility and will match other containers naming

**Comment by jonorossi** [comment]
@generik0 I'd avoid using .NET Core branding since Core is slowly disappearing, the project is named ".NET Extensions".

@ltines let's go for it, no real reason to keep the old namespace convention since we've deprecated so many facility in recent years anyway.

Only nitpick with `Castle.Windsor.Extensions.MsDependencyInjection`, is that the "Extensions" bit now with the MS prefix isn't referring to the "Microsoft.Extensions" (.NET Extensions) project, and everything not in `Castle.Windsor` really is an extension, so `Castle.Windsor.MsDependencyInjection` would be enough, or do as Autofac did `Castle.Windsor.Extensions.DependencyInjection`.

So:
- `Castle.Windsor.Extensions.DependencyInjection`, or
- `Castle.Windsor.ExtensionsDependencyInjection`, or
- `Castle.Windsor.MsDependencyInjection`

With my preference now to 1 or 2, since the project is ".NET Extensions DependencyInjection".

**Comment by robsonj** [comment]
`Castle.Windsor.MsDependencyInjection` - would make the most sense to me given that castle windsor itself is a dependency injection framework, without the `Ms` It could be confusing as to what that namespace is the subject of.

**Comment by generik0** [comment]
@ltines i have written to you on issues for your own repo, to initialize some planing and your expectations. 
Please tell me how you would like me to assist.
I can start by integrating what you have already done on Friday?

**Comment by jonorossi** [comment]
I also just remembered that `Castle.Windsor.MsDependencyInjection` would conflict with the existing volosoft package with the same name.

**Comment by ltines** [comment]
@generik0 i'll move the code as PR here (once i figure out how to do it)

@jonorossi @robsonj 
`Castle.Windsor.MsDependencyInjection` would be beautfiul. So i still think the closes is `Castle.Windsor.Extensions.MsDependencyInjection`. It's an extension to .NET Core

**Comment by generik0** [comment]
@ltines that's perfect! Let me know when you have a pre-release or something ready. FYI, The castle project builds a little funny (to a build folder), so from what I remember your dll/nugets land there (just a hint).

Regarding name. I vote for "Castle.Windsor.Extensions.DependencyInjection", Don't like the MS prefix, as you will need to change the name if MS changed the name of there repo, (or gave the whole .net core away to a foundation, like Oracle did for apache)... Thats just my thoughts. Do as you feel best @ltines 

Either way, looks like I can just sit back and play the tester in a bit, and that is nerver a bad role :-).
Let me know when you have a fork ready with the code please  @ltines.

Looking forward to hearing more. Great work everyone. Great work @ltines :-)

**Comment by ltines** [comment]
Any update guys?

**Comment by jonorossi** [comment]
> Any update guys?

Sorry for not getting to this yet, been busy but will make time on Sunday.

**Comment by robertmircea** [comment]
In which state is the nuget package? Was it published?

**Comment by jonorossi** [comment]
I was last to comment on #517 10 days ago, not sure where @ltines is at.

**Comment by jonorossi** [comment]
Published v5.1.0-beta001:
https://www.nuget.org/packages/Castle.Windsor.Extensions.DependencyInjection/

Please log any problems in new issues.

Thanks everyone.

**Comment by robsonj** [comment]
5.1.0-beta001 working fine for me. Is there any thought on an approximate release timeframe, i.e 1week, 1month, 1quarter away?

**Comment by generik0** [comment]
I have a rather large aspnetcore project with many integrations (e.g. including node services). I would like to try the beta.
Just need a little time to get time and do a update prototype...

Has anyone tested the EF Core integration security integration with the package?

**Comment by chrisheil** [comment]
Great work! Thank you all for putting in the hard work to get this added.

One question, what would the proper setup be for this update? [The documentation](https://github.com/castleproject/Windsor/blob/master/docs/aspnetcore-facility.md) still mentions using the old `Castle.Facilities.AspNetCore` package

**Comment by generik0** [comment]
@chrisheil please look here:

https://github.com/castleproject/Windsor/blob/master/docs/net-dependency-extension.md

@ltines did a great job. The aspnetcore facility has been improved with the net dependence extension for dotnet+

**Comment by jonorossi** [comment]
> 5.1.0-beta001 working fine for me. Is there any thought on an approximate release timeframe, i.e 1week, 1month, 1quarter away?

@robsonj All depends on when you all are happy for it to go out after bugs are ironed out. So the more use now will likely speed up people's confidence.

> One question, what would the proper setup be for this update? [The documentation](https://github.com/castleproject/Windsor/blob/master/docs/aspnetcore-facility.md) still mentions using the old `Castle.Facilities.AspNetCore` package

@chrisheil @generik0 I've just edited the warning on the old ASP.NET Core 2.x facility page. Feel free to send pull requests to improve the documentation.

**Comment by chrisheil** [comment]
@generik0 That did the trick, thank you.

I have been running this in our test servers for a couple days and experiencing no issues. I did have to find a workaround for replacing the Volosoft `MsScopedLifestyleManager` with `ScopedToNetServiceScope` on one of my registrations. In case anyone else runs into a similar problem here's how I worked around it.

Original code:
``` csharp
container.Register(Classes
    .FromAssemblyContaining<IFoo>()
    .BasedOn<IFoo>()
    .WithServiceBase()
    .LifestyleCustom<MsScopedLifestyleManager>()); // From the Volosoft nuget package
```

Updated code
``` csharp
container.Register(Classes
    .FromAssemblyContaining<IFoo>()
    .BasedOn<IFoo>()
    .WithServiceBase() // <== Cannot use ".Lifestyle.ScopedToNetServiceScope()"
    .Configure(component => component.LifeStyle.ScopedToNetServiceScope()));
```

**Comment by robertmircea** [comment]
I am trying to inject `IHttpClientFactory` as a constructor parameter to one of my classes and I receive the following exception when using `Castle.Windsor.Extensions.DependencyInjection`

`DefaultHttpClientFactory` is an internal class which is registered by Microsoft for supporting `IHttpClientFactory`.

Previously, I was using Volosoft with no issues.

Any idea on how to solve this?

```
Castle.MicroKernel.ComponentNotFoundException: No component for supporting the service Microsoft.Extensions.Http.DefaultHttpClientFactory was found
   at Castle.MicroKernel.DefaultKernel.Castle.MicroKernel.IKernelInternal.Resolve(Type service, Arguments arguments, IReleasePolicy policy, Boolean ignoreParentContext)
   at Castle.MicroKernel.DefaultKernel.Resolve(Type service, Arguments arguments)
   at Castle.Windsor.WindsorContainer.Resolve(Type service)
   at Castle.Windsor.Extensions.DependencyInjection.WindsorScopedServiceProvider.ResolveInstanceOrNull(Type serviceType, Boolean isOptional)
   at Castle.Windsor.Extensions.DependencyInjection.WindsorScopedServiceProvider.GetRequiredService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.HttpClientFactoryServiceCollectionExtensions.<>c.<AddHttpClient>b__0_0(IServiceProvider serviceProvider)
   at Castle.Windsor.Extensions.DependencyInjection.RegistrationAdapter.<>c__DisplayClass4_0`1.<UsingFactoryMethod>b__0(IKernel kernel)
   at Castle.MicroKernel.Registration.ComponentRegistration`1.<>c__DisplayClass84_0`1.<UsingFactoryMethod>b__0(IKernel k, ComponentModel m, CreationContext c)
   at Castle.MicroKernel.ComponentActivator.FactoryMethodActivator`1.Instantiate(CreationContext context)
   at Castle.MicroKernel.ComponentActivator.DefaultComponentActivator.InternalCreate(CreationContext context)
   at Castle.MicroKernel.ComponentActivator.AbstractComponentActivator.Create(CreationContext context, Burden burden)
   at Castle.MicroKernel.Lifestyle.AbstractLifestyleManager.CreateInstance(CreationContext context, Boolean trackedExternally)
   at Castle.MicroKernel.Lifestyle.ScopedLifestyleManager.<>n__0(CreationContext context, Boolean trackedExternally)
   at Castle.MicroKernel.Lifestyle.ScopedLifestyleManager.<>c__DisplayClass4_0.<Resolve>b__0(Action`1 afterCreated)
   at Castle.Windsor.Extensions.DependencyInjection.Scope.ExtensionContainerScope.GetCachedInstance(ComponentModel model, ScopedInstanceActivationCallback createInstance)
   at Castle.MicroKernel.Lifestyle.ScopedLifestyleManager.Resolve(CreationContext context, IReleasePolicy releasePolicy)
   at Castle.MicroKernel.Handlers.DefaultHandler.ResolveCore(CreationContext context, Boolean requiresDecommission, Boolean instanceRequired, Burden& burden)
   at Castle.MicroKernel.Handlers.DefaultHandler.Resolve(CreationContext context, Boolean instanceRequired)
   at Castle.MicroKernel.Handlers.AbstractHandler.Resolve(CreationContext context)
   at Castle.MicroKernel.Resolvers.DefaultDependencyResolver.ResolveFromKernelByType(CreationContext context, ComponentModel model, DependencyModel dependency)
   at Castle.MicroKernel.Resolvers.DefaultDependencyResolver.ResolveFromKernel(CreationContext context, ComponentModel model, DependencyModel dependency)
   at Castle.MicroKernel.Resolvers.DefaultDependencyResolver.TryResolveCore(CreationContext context, ISubDependencyResolver contextHandlerResolver, ComponentModel model, DependencyModel dependency, Object& value)
   at Castle.MicroKernel.Resolvers.DefaultDependencyResolver.Resolve(CreationContext context, ISubDependencyResolver contextHandlerResolver, ComponentModel model, DependencyModel dependency)
   at Castle.MicroKernel.ComponentActivator.DefaultComponentActivator.CreateConstructorArguments(ConstructorCandidate constructor, CreationContext context)
   at Castle.MicroKernel.ComponentActivator.DefaultComponentActivator.Instantiate(CreationContext context)
   at Castle.MicroKernel.ComponentActivator.DefaultComponentActivator.InternalCreate(CreationContext context)
   at Castle.MicroKernel.ComponentActivator.AbstractComponentActivator.Create(CreationContext context, Burden burden)
   at Castle.MicroKernel.Lifestyle.AbstractLifestyleManager.CreateInstance(CreationContext context, Boolean trackedExternally)
   at Castle.MicroKernel.Lifestyle.SingletonLifestyleManager.Resolve(CreationContext context, IReleasePolicy releasePolicy)
   at Castle.MicroKernel.Handlers.DefaultHandler.ResolveCore(CreationContext context, Boolean requiresDecommission, Boolean instanceRequired, Burden& burden)
   at Castle.MicroKernel.Handlers.DefaultHandler.Resolve(CreationContext context, Boolean instanceRequired)
   at Castle.MicroKernel.Handlers.AbstractHandler.Resolve(CreationContext context)
   at Castle.MicroKernel.Resolvers.DefaultDependencyResolver.ResolveFromKernelByType(CreationContext context, ComponentModel model, DependencyModel dependency)
   at Castle.MicroKernel.Resolvers.DefaultDependencyResolver.ResolveFromKernel(CreationContext context, ComponentModel model, DependencyModel dependency)
   at Castle.MicroKernel.Resolvers.DefaultDependencyResolver.TryResolveCore(CreationContext context, ISubDependencyResolver contextHandlerResolver, ComponentModel model, DependencyModel dependency, Object& value)
   at Castle.MicroKernel.Resolvers.DefaultDependencyResolver.Resolve(CreationContext context, ISubDependencyResolver contextHandlerResolver, ComponentModel model, DependencyModel dependency)
   at Castle.MicroKernel.ComponentActivator.DefaultComponentActivator.CreateConstructorArguments(ConstructorCandidate constructor, CreationContext context)
   at Castle.MicroKernel.ComponentActivator.DefaultComponentActivator.Instantiate(CreationContext context)
   at Castle.MicroKernel.ComponentActivator.DefaultComponentActivator.InternalCreate(CreationContext context)
   at Castle.MicroKernel.ComponentActivator.AbstractComponentActivator.Create(CreationContext context, Burden burden)
   at Castle.MicroKernel.Lifestyle.AbstractLifestyleManager.CreateInstance(CreationContext context, Boolean trackedExternally)
   at Castle.MicroKernel.Lifestyle.SingletonLifestyleManager.Resolve(CreationContext context, IReleasePolicy releasePolicy)
   at Castle.MicroKernel.Handlers.DefaultHandler.ResolveCore(CreationContext context, Boolean requiresDecommission, Boolean instanceRequired, Burden& burden)
   at Castle.MicroKernel.Handlers.DefaultHandler.Resolve(CreationContext context, Boolean instanceRequired)
   at Castle.MicroKernel.Handlers.AbstractHandler.Resolve(CreationContext context)
   at Castle.MicroKernel.Resolvers.DefaultDependencyResolver.ResolveFromKernelByType(CreationContext context, ComponentModel model, DependencyModel dependency)
   at Castle.MicroKernel.Resolvers.DefaultDependencyResolver.ResolveFromKernel(CreationContext context, ComponentModel model, DependencyModel dependency)
   at Castle.MicroKernel.Resolvers.DefaultDependencyResolver.TryResolveCore(CreationContext context, ISubDependencyResolver contextHandlerResolver, ComponentModel model, DependencyModel dependency, Object& value)
   at Castle.MicroKernel.Resolvers.DefaultDependencyResolver.Resolve(CreationContext context, ISubDependencyResolver contextHandlerResolver, ComponentModel model, DependencyModel dependency)
   at Castle.MicroKernel.ComponentActivator.DefaultComponentActivator.CreateConstructorArguments(ConstructorCandidate constructor, CreationContext context)
   at Castle.MicroKernel.ComponentActivator.DefaultComponentActivator.Instantiate(CreationContext context)
   at Castle.MicroKernel.ComponentActivator.DefaultComponentActivator.InternalCreate(CreationContext context)
   at Castle.MicroKernel.ComponentActivator.AbstractComponentActivator.Create(CreationContext context, Burden burden)
   at Castle.MicroKernel.Lifestyle.AbstractLifestyleManager.CreateInstance(CreationContext context, Boolean trackedExternally)
   at Castle.MicroKernel.Lifestyle.SingletonLifestyleManager.Resolve(CreationContext context, IReleasePolicy releasePolicy)
   at Castle.MicroKernel.Handlers.DefaultHandler.ResolveCore(CreationContext context, Boolean requiresDecommission, Boolean instanceRequired, Burden& burden)
   at Castle.MicroKernel.Handlers.DefaultHandler.Resolve(CreationContext context
```

**Comment by generik0** [comment]
Hi @robertmircea 
Did you register it?

**Comment by robertmircea** [comment]
I didn't have to register it when using Volosoft's adapter. `DefaultHttpClientFactory` was automatically registered when calling `services.AddHttpClient()` in `Startup.cs` to register my http clients. However, after switching to `Castle.Windsor.Extensions.DependencyInjection`, the same code (no modifications) no longer works.

**Comment by generik0** [comment]
I see, so you have services.AddHttpClient() which should register it, but just to the castle container.
Did you use the other package with net core 3 also? Because it might be changes in net core 3?

Maybe the default factory needs to be registered first with “IncludeNonPublicTypes”?

Maybe find the code for the default factory aspnetcore lib? Also check what the “AddHttpClient” is doing in the github repo, and try doing it yourself with castle registrations?

**Comment by robsonj** [comment]
I was wondering what the current thoughts are on non beta Castle release to include this feature?

**Comment by jonorossi** [comment]
@robsonj there are a couple of new issues open, no idea if they are blockers.

**Comment by generik0** [comment]
Hi @ltines, this might be a stupid question. But the "WebHost" does not allow the extension to register castle Windsor as the service provider- only "Host".
Any thoughts? Shouldn't there be an extension to WebHost?

**Comment by generik0** [comment]
@ltines ie:
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/web-host?view=aspnetcore-3.1

![image](https://user-images.githubusercontent.com/2411563/86596941-52e81900-bf9b-11ea-80a2-8937893b8bde.png)

**Comment by jonorossi** [comment]
@generik0 It is best to create new issues to track things since this is already closed and released (as a beta).

**Comment by rmszc81** [comment]
hello everyone,
is there any date for release the final version of this component? for now it's still in beta, :(
thanks!

**Comment by generik0** [comment]
Hi @rmszc81 have you been using the Beta?
We are almost through the issues We know about, but still need more beta testers to make sure all configurations are Working

**Comment by rmszc81** [comment]
Yeah, I have been using the beta and it's working fine for me, so, in short:

it's an asp.net core 3.1 web api application, including:
- automapper
- serilog
- castle dynamic proxy,
- castle windsor (obviously)

injections are mostly done by constructors, a few by properties.

no known issues, it's running properly and performance is good too.

**Comment by chrisheil** [comment]
@generik0 We upgraded our application the day the beta came out, and have been cleared in QA and UAT after several months of testing. In fact, this beta resolved some issues we had faced with the workarounds we had introduced with the earlier solutions before the beta

We are running on asp.netcore 3.0 web api with the following:

* automapper
* castle dynamic proxy
* graylog
* Entity Framework Core 2.2
* FluentValidation

We are also upgrading another product to use this beta with similar dependencies, and early signs are positive

**Comment by generik0** [comment]
Good news @chrisheil. Thanks!!!!
